docs/proposals/diamond-design.md

```diff
@@ -0,0 +1,458 @@
+# Diamond workflow support
+
+Diamond workflows allow for a pipeline to scale out the generation of a single bundle.
+
+A diamond is composed of multiple splits; each split tries to write a set of files. Ideally, 2 splits do not write the
+same file.
+
+To elaborate, a bundle consists of a filesystem namespace with files in folders; each split can write files to any
+folder but should not try to write the same file.
+
+The individual nodes add files to splits without any co-ordination between them.
+
+Commit logic handles any conflicts that exist. Conflicts can only occur if the same file in the namespace of the bundle is written to by more than one split with different data.
+
+Diamonds and splits are essentially transitory objects, which live in the metadata namespace of a datamon repo.
+
+
+## Workflow
+
+The diamond workflow consists of the following steps.
+
+1. For a diamond to start, datamon needs some unique ID. Datamon SDK allows for the generation of such a unique ID for a diamond.
+
+   1. Prerequisite: generate a unique KSUID and a signalling metadata object that allows for starting splits.
+   2. All clients participating to the same diamond are expected to use the same ID at each step. **This is a requirement on client coordination**
+   3. The {diamond-id} is only used once: users cannot initialize the same diamond twice.
+   4. Initialization command, returning an identifier which is guaranteed to be unique:
+      ```
+      datamon bundle diamond initialize --repo {repo}
+      {diamond-id}
+      ```
+   5. Backend model: the following path is created in the vmetadata bucket.
+      ```
+      /splits/{repo}/{diamond-id}/diamond.yaml
+      ```
+
+> **Trade-off**: this requirement is set on purpose. It is not guided by implementation constraints but by the desire to protect users against
+> unintentional, silent data corruption to could occur if the diamond ID is left as a user-controlled parameter.
+
+2. Uploading files in a diamond split stores files as blobs, then generates a file list in the versioned metadata bucket.
+
+   1. Files are uploaded by clients concurrently. This operation does not result in a new bundle yet.
+      Files are in a pre-commited state, meaning that all content is persisted as blobs on the CAFS store.
+   2. Command:
+      ```
+      datamon bundle diamond add --repo {repo} --diamond {diamond-id} --path {source}
```

Comment:
one downside of not having ```split-id``` is that we not create conflicting files when a client retires the same files due to failure.. so if the customer wants to set commit with no conflict and there is no real conflict as the splits wrote independent files we would artificially introduce conflicts, breaking their failure handling and retry mechanism for a pod in the diamond. I think we will need to reintroduce the ```split-id``` or make it an optional flag where we honor the split-id and not internally generate one.

Reply:
Also adding the suffix for the split id to the conflicting files allows the developer to better understand the issue with their logic, we could add our own internal KSUID but that is harder to use for the developer.

Reply:
Retrying same files without changing their content shall not trigger a conflict.

Reply:
Retrying same files without changing their content shall not trigger a conflict.

Reply:
An interesting case i haven t checked is when you remove a file from the dataset. Todo

Reply:
now split ID may be injected to cover the discussed use-case (now detailed).
didn't provide details about removing files, although providing explicit split ID should cover that case.

Reply:
explicit split id supported